docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1017,6 +1019,129 @@ table my_ipv4_lkup {
 }
 ~ End p4example
 
+### Custom Annotations in P4Info
+P4 allows flexible annotations[@P4Annotations]. Users may add their own
+annotations as long as they obey the language syntax. To facilitate consumption
+of user-defined annotations, P4Info has an `AnnotationMap` for any element which
+can accept such annotations, &ie; the `PkgInfo`, `Preamble`, `MatchField`,
+`ActionRef`, `Action` and `ControllerPacketMetadata` messages, as well as the
+type specification messages in `p4types.proto`. The `AnnotationMap` contains a
+suitable parsed representation for the different types of annotation bodies,
+namely: `expressions` and `properties`. Empty annotation bodies are signified by
+an empty `value` field  in the `AnnotationMap` entry. 
+
+Note that `AnnotationBody` allows for simultaneous `expressions` and
+`properties` entries even though the P6-16 specification grammar does not at this
+writing.
+
+The `AnnotationBody` and `AnnotationMap` messages are defined in `p4types.proto`:
+
+~ Begin Proto
+// A parsed representation of the annotation body.
+// Ex. empty annotation:           @my_anno1()
+// Ex. expression list annotation: @my_anno2(1+1, "Hello", 1 << 3)
+// Ex. KV-pair annotation:         @my_anno3(a=b,c=d)
+// Ex. Cumulative kv-pair          @my_anno3(e=f)  // adds to my_anno3's properties
+message AnnotationBody {
+  repeated string expressions = 1;
```

Comment:
Should this be a oneof?  I think it can only be one or the other, never both, correct?

Reply:
@stefanheule , Originally in my proposal it was `oneof`, but later I removed `oneof` to allow for future enhancement to the P4-16 spec.If you look at the `@my_anno5` example I cited, I can value in allowing a mix of kv-pairs and expressions

Reply:
Ah my bad, I should have followed the previous discussion more closely.  But my understanding is correct that P4-16 spec right now only allows one or the other, correct?  Are we aware of plans that they want to generalize this?

We might not want to have an example of something that isn't actually valid according to the current spec (even if we decide not to use a oneof).

Reply:
Just to close this thread, I was actually wrong, the P4-16 spec does (sort of) allow mixing them, by accumulating through the use of multiple annotations.

> Multiple annotations of either form may be used and are cumulative, i.e. the list of k-v pairs will accumulate.

To me, this reads that `@Foo(1) @Foo(a=b)` is allowed, and will result in (conceptually) a single annotation `Foo` that has both `1` and `a=b` as it's elements.  Writing it as a single annotation (as in `@Foo(1, a=b)`) is forbidden by the grammer, though, not sure if that's on oversight or on purpose.

Reply:
Thinking about this some more, this is at least my interpretation of what the P4 spec says.  However, we may want to ask if that's correct, and why there is this inconsistency between accumulating expression lists and KV lists, but not allowing mixing of them directly.

Another thing we might want to check is if the compiler currently accumulates/overwrites as is described in the P4 spec.  If it does not, we should ask why.  If it does, a P4RT implementation could reuse that when creating the proto.